clang/lib/DPCT/BarrierFenceSpaceAnalyzer.h

```diff
@@ -82,13 +112,113 @@ class BarrierFenceSpaceAnalyzer
   // FDLoc is in the map means this kernel function is analyzed.
   // CELoc is not in the map means cannot set local fence space.
   void setFalseForThisFunctionDecl() {
-    CachedResults[FDLoc] = std::unordered_map<std::string, bool>();
+    if (!IsDryRun)
+      CachedResults[FDLoc] =
+          std::unordered_map<std::string, BarrierFenceSpaceAnalyzerResult>();
   }
 
   /// (FD location, (Call location, result))
-  static std::unordered_map<std::string, std::unordered_map<std::string, bool>>
+  static std::unordered_map<
+      std::string,
+      std::unordered_map<std::string, BarrierFenceSpaceAnalyzerResult>>
       CachedResults;
-  static const std::unordered_set<std::string> AllowedDeviceFunctions;
+  bool IsDryRun = false;
+  bool MayDependOn1DKernel = false;
+
+  template <class TargetTy, class NodeTy>
+  static inline const TargetTy *findAncestorInFunctionScope(
+      const NodeTy *N, const FunctionDecl *Scope,
+      const std::function<const void *(const DynTypedNode &,
+                                       const DynTypedNode &)> &Operation) {
+    auto &Context = DpctGlobalInfo::getContext();
+    DynTypedNode Current = DynTypedNode::create(*N);
+    DynTypedNodeList Parents = Context.getParents(Current);
+    while (!Parents.empty()) {
+      if (Parents[0].get<FunctionDecl>() &&
+          Parents[0].get<FunctionDecl>() == Scope)
+        break;
+      if (const void *Node = Operation(Parents[0], Current)) {
+        return reinterpret_cast<const TargetTy *>(Node);
+      }
+      Current = Parents[0];
+      Parents = Context.getParents(Current);
+    }
+    return nullptr;
+  }
+
+  bool isPotentialGlobalMemoryAccess(std::shared_ptr<DeviceFunctionInfo> DFI,
+                                     bool IsInGlobalFunction);
+  std::set<const Expr *> DeviceFunctionCallArgs;
+
+  class TypeAnalyzer {
+  public:
+    /// Decide the input parameter type class
+    /// \return One of below 3 int values:
+    ///  1: can skip analysis
+    ///  0: need analysis
+    /// -1: unsupport to analyze
+    int getInputParamterTypeKind(clang::QualType QT) {
+      bool Res = getTypeInfo(QT.getTypePtr());
+      if (!Res)
+        return -1;
```

Comment:
using Enum value instead of magic number to improve readability.

Reply:
done